tools/qtscript/qtscript.man
.TH "qtscript" "1" "Jolla Ltd." "Jan 30, 2013"
.SH "NAME"
.PP
qtscript allows to run JavaScript and QML files with QtScript support
.SH "SYNOPSIS"
.PP
\fBqtscript\fP [-h] \fIscript\d\s-2file\s+2\u\fP [script arguments]
.SH "DESCRIPTION"
.PP
QtScript is the Qt library JavaScript engine. qtscript command
provides the way to run JavaScript files as shell scripts or to load QML
UI with QtScript support.
.SH "COMMAND LINE OPTIONS"
.TP
\fB\fB-h\fP\fP
show usage
.TP
\fB\fBscript\d\s-2file\s+2\u\fP\fP
name of the javascript or qml file to be executed,
files with .qml extension will be considered as 
QML UI files, other - as JS files
.TP
\fB\fB[script arguments]\fP\fP
arguments to be passed to javascript in
the \fBqtscript.args\fP variable
.SH "ENVIRONMENT"
.TP
\fB\fBQTSCRIPT\_LIBRARY\_PATH\fP\fP
contains colon-separated list of
paths to directories containing QtScript extensions and
JavaScript modules, available for importing. Because this paths
are just library paths, extensions themselves (shared objects)
should reside in subdirectories ./script/. JavaScript modules
should reside in ./. By default qtscript is trying to add
/usr/lib{,32,64}/qt4/plugins directories.
.SH "JAVASCRIPT CONTEXT"
.PP
qtscript provides several objects/functions, available in executed
scripts. All this variables are members of the global \fBqtscript\fP
object.

.TP
\fB\fBqtscript.system.os\fP\fP
OS "name": can be any of ["linux", "macos",
"win32", "unknown"]
.TP
\fB\fBqtscript.system.env\fP\fP
environment variables
.TP
\fB\fBqtscript.system.path\fP\fP
Application library paths
.TP
\fB\fBqtscript.load\fP\fP
function to load javascript file located in the
directory where running script resides or in the library path
.TP
\fB\fBqtscript.use\fP\fP
instructs engine to import QtScript extension
.TP
\fB\fBqtscript.script.args\fP\fP
list of command line arguments
.TP
\fB\fBqtscript.script.filename\fP\fP
currently executed script file name
(e.g. in \fBqtscript.eval\fP it will be evaluated script path name)
.TP
\fB\fBqtscript.script.cwd\fP\fP
current working directory or directory where
executed/evaluated script is located
.SH "AUTHOR"
.PP
Denis Zalevskiy <denis.zalevskiy@jollamobile.com>
.SH "LICENSING"
.PP
qtscript is distributed under terms and conditions of the LGPLv2
license.


qtscript.cpp
#include <QCoreApplication>
#include <QtScript>
#include <QStringList>
#include <QScopedPointer>
#include <QFile>
#include <QTextStream>
#include <QProcess>
#include <QtDebug>

#include <iostream>
#include <stdexcept>

class Error : public std::runtime_error
{
public:
    Error(QString const &s)
        : std::runtime_error(s.toStdString()),
          msg(s)
    { }

    virtual ~Error() throw() {}

    QString msg;
};

class JsError : public Error
{
public:
    static QString errorMessage(QScriptEngine &engine, QString const &file)
    {
        QString res;
        QTextStream stream(&res);
        stream << "Exception: "
               << engine.uncaughtException().toString() << "\n"
               << file << ":" << engine.uncaughtExceptionLineNumber()
               << engine.uncaughtExceptionBacktrace().join("\n")
               << "\n";
        return res;

    }

    JsError(QScriptEngine &engine, QString const &file)
        : Error(errorMessage(engine, file)) {}
};

#ifdef Q_OS_WIN32
static const char *os_name = "windows";
#elif defined(Q_OS_LINUX)
static const char *os_name = "linux";
#elif defined(Q_OS_MAC)
static const char *os_name = "macos";
#elif defined(Q_OS_UNIX)
static const char *os_name = "unix";
#else
static const char *os_name = "unknown";
#endif

struct ScriptValueAccessor
{
    ScriptValueAccessor(QString const &name, QScriptValue v)
        : name(name), value(v) {}

    QString name;
    QScriptValue value;
};

ScriptValueAccessor operator <<
(ScriptValueAccessor to, ScriptValueAccessor const& child)
{
    to.value.setProperty(child.name, child.value);
    return to;
}

QScriptValue& operator <<
(QScriptValue &to, ScriptValueAccessor const& child)
{
    to.setProperty(child.name, child.value);
    return to;
}

static ScriptValueAccessor nameValue(QString const &n, QScriptValue v)
{
    return ScriptValueAccessor(n, v);
}

static QMap<QString, QScriptValue> saveProps
(QScriptValue const &parent, QStringList const &names)
{
    QMap<QString, QScriptValue> res;
    for (auto &name : names)
        res[name] = parent.property(name);
    return res;
}

static void restoreProps
(QScriptValue &parent, QMap<QString, QScriptValue> const& props)
{
    for (auto kv = props.begin(); kv != props.end(); ++kv)
        parent.setProperty(kv.key(), kv.value());
}

static QScriptValue globalProp
(QScriptEngine const& engine, QStringList const &path)
{
    auto res = engine.globalObject();
    for (auto &name : path)
        res = res.property(name);

    return res;
}

static int usage(int, char *argv[])
{
    qDebug() << argv[0] << " <script_name>";
    return 0;
}

static QScriptValue eval(QString file_name, QScriptEngine &engine)
{
    static QSet<QString> loaded_files;

    QFileInfo file_info(file_name);
    file_name = file_info.absoluteFilePath();

    QString canonical_name = file_info.canonicalFilePath();
    if (loaded_files.contains(canonical_name))
        return QScriptValue();

    QFile file(file_name);
    if (!file.open(QFile::ReadOnly))
        throw Error(QString("Can't open %1").arg(file_name));

    QString contents;
    contents.reserve(file_info.size());

    int line_nr = 2;

    QTextStream dst(&contents);
    QTextStream input(&file);
    QString first = input.readLine();
    if (!first.startsWith("#!")) {
        dst << first << "\n";
        line_nr = 1;
    }

    while (!input.atEnd())
        dst << input.readLine() << "\n";

    auto script = globalProp(engine, {"app", "script"});

    auto saved_props = std::move(saveProps(script, {"filename", "cwd"}));
    script << nameValue("filename", engine.toScriptValue(file_name))
           << nameValue("cwd", engine.toScriptValue(file_info.absolutePath()));

    auto res = engine.evaluate(contents, file_name, line_nr);
    restoreProps(script, saved_props);

    if (engine.hasUncaughtException())
        throw JsError(engine, file_name);

    loaded_files.insert(canonical_name);
    return res;
}

static QList<QDir> lib_dirs;

static QString findFile(QScriptEngine &engine, QString const &file_name)
{
    auto script = globalProp(engine, {"app", "script"});
    QDir cwd(script.property("cwd").toString());
    QString res;

    if (QFileInfo(file_name).isRelative())
        res = cwd.filePath(file_name);

    if (QFileInfo(res).exists())
        return res;

    for (auto &d : lib_dirs) {
        res = d.filePath(file_name);
        if (QFileInfo(res).exists())
            return res;
    }
    return QString();
}

static QScriptValue jsEval(QScriptContext *context, QScriptEngine *engine)
{
    QString file_name = context->argument(0).toString();
    try {
        QString full_name = findFile(*engine, file_name);
        if (full_name.isEmpty())
            throw Error(QString("Can't find file %1").arg(file_name));
        return eval(full_name, *engine);
    } catch (JsError const &e) {
        return context->throwError
            (QString("Exception loading file %1").arg(file_name) + e.msg);
    } catch (Error const &e) {
        return context->throwError
            (QString("Exception loading file %1:\n%2").arg(file_name, e.msg));
    }
}

static QScriptValue jsUse(QScriptContext *context, QScriptEngine *engine)
{
    return engine->importExtension(context->argument(0).toString());
}

static QMap<QString, QVariant> mk_env()
{
    QMap<QString, QVariant> res;
    QStringList env = QProcess::systemEnvironment();
    for (auto &item : env) {
        QStringList kv = item.split('=');
        auto s = kv.size();
        if (s)
            res.insert(kv[0], (s == 1) ? QString() : kv[1]);
    }
    return res;
}

int main(int argc, char *argv[])
{
    QCoreApplication app(argc, argv);
    if (app.arguments().size() < 2)
        return usage(argc, argv);

    QString script_file(app.arguments().at(1));

    auto env = std::move(mk_env());

    auto lib_paths = env["QTSCRIPT_LIBRARY_PATH"].toString().split(":");
    for (auto path : {"/usr/lib/qt4/plugins", "/usr/lib32/qt4/plugins",
                "/usr/lib64/qt4/plugins"}) {
        if (QDir(path).exists())
            lib_paths.push_back(path);
    }

    for (auto &path : lib_paths)
        lib_dirs.push_back(QDir(path));
    app.setLibraryPaths(lib_paths);

    QScriptEngine engine;

    auto obj = [&engine](QString const &name)
        { return nameValue(name, engine.newObject()); };

    auto fn = [&engine]
        (QString const &name, QScriptEngine::FunctionSignature sig)
        { return nameValue(name, engine.newFunction(sig)); };

    auto global = engine.globalObject();
    global << (obj("app")
               << (obj("system")
                   << nameValue("os", engine.toScriptValue(QString(os_name)))
                   << nameValue("env", engine.toScriptValue(env))
                   << nameValue("path", engine.toScriptValue(lib_paths)))
               << (obj("script")
                   << fn("eval", jsEval)
                   << fn("use", jsUse)
                   << nameValue("args", engine.toScriptValue(app.arguments()))))
           << obj("lib");

    try {
        auto res = eval(script_file, engine);
        std::cout << res.toString().toStdString();
    } catch (Error const &e) {
        qDebug() << "Failed to eval:" << script_file;
        qDebug() << e.msg;
        return -1;
    }

    return 0;
}
